source/base/tensor_function_parser.cc

```diff
@@ -0,0 +1,419 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2005 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/tensor_function_parser.h>
+#include <deal.II/base/mu_parser_internal.h>
+#include <deal.II/base/patterns.h>
+#include <deal.II/base/thread_management.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/base/tensor.h>
+#include <deal.II/base/tensor_function.h>
+
+#include <boost/random.hpp>
+
+#include <cmath>
+#include <map>
+
+#ifdef DEAL_II_WITH_MUPARSER
+#  include <muParser.h>
+#else
+
+
+
+namespace fparser
+{
+  class TensorFunctionParser
+  {};
+} // namespace fparser
+#endif
+
+DEAL_II_NAMESPACE_OPEN
+
+
+template <int rank, int dim, typename Number>
+const std::vector<std::string> &
+TensorFunctionParser<rank, dim, Number>::get_expressions() const
+{
+  return expressions;
+}
+
+
+
+template <int rank, int dim, typename Number>
+TensorFunctionParser<rank, dim, Number>::TensorFunctionParser(const double initial_time)
+  : TensorFunction<rank, dim, Number>(initial_time)
+  , initialized(false)
+  , n_vars(0)
+  , n_components(std::pow(dim,rank))
+{}
+
+
+template <int rank, int dim, typename Number>
+TensorFunctionParser<rank, dim, Number>::TensorFunctionParser(const std::string &expression,
+														const std::string &constants,
+														const std::string &variable_names)
+  : TensorFunction<rank, dim, Number>()
+  , initialized(false)
+  , n_vars(0)
+  , n_components(std::pow(dim,rank))
+{
+  auto constants_map = Patterns::Tools::Convert<ConstMap>::to_value(
+    constants,
+    std_cxx14::make_unique<Patterns::Map>(Patterns::Anything(),
+                                          Patterns::Double(),
+                                          0,
+                                          Patterns::Map::max_int_value,
+                                          ",",
+                                          "="));
+  initialize(variable_names,
+             expression,
+             constants_map,
+             Utilities::split_string_list(variable_names, ",").size() ==
+               dim + 1);
+}
+
+
+// We deliberately delay the definition of the default destructor
+// so that we don't need to include the definition of mu::Parser
+// in the header file.
+template <int rank, int dim, typename Number>
+TensorFunctionParser<rank, dim, Number>::~TensorFunctionParser() = default;
```

Comment:
Why can't we just use the default destructor from the beginning (i.e., don't declare it at all)? I don't understand why this is necessary.

Reply:
I left it as is since I saw that you opened another issue #9143